cpp/src/arrow/dataset/filter.cc

```diff
@@ -715,34 +707,46 @@ std::string OperatorName(compute::CompareOperator op) {
   return "";
 }
 
+// TODO(bkietz) extract this to Scalar::ToString()
+struct ScalarExpressionToString {
+  Status Visit(const BooleanType&) {
+    return Finish(CastValue<BooleanType>().value ? "true" : "false");
+  }
+
+  template <typename T>
+  enable_if_number<T, Status> Visit(const T&) {
+    return Finish(std::to_string(CastValue<T>().value));
+  }
+
+  Status Visit(const StringType&) {
+    return Finish(CastValue<StringType>().value->ToString());
+  }
+
+  Status Visit(const DataType&) { return Finish("TODO(bkietz)"); }
```

Comment:
Return `Status::NoImplemented`.

Reply:
This is used in non-failable `std::string ToString()` so I DCHECK the result of this visitor. Returning `NotImplemented` will abort if ToString() is called on a scalar expression which I haven't provided print method for, is that better?